I've also tested this branch on my environment (Ubuntu 22.04, WSL2).

It ends with same error:

<details>

<summary>test lz4 qemu-m68k-static platform</summary>

```
 ---- test lz4 qemu-m68k-static platform ----
qemu-m68k-static ./datagen -g16KB  | qemu-m68k-static ../programs/lz4 -9     | qemu-m68k-static ../programs/lz4 -t

stdin                          : decoded 16384 bytes
qemu-m68k-static ./datagen         | qemu-m68k-static ../programs/lz4        | qemu-m68k-static ../programs/lz4 -t

stdin                          : decoded 65536 bytes
qemu-m68k-static ./datagen -g256MB | qemu-m68k-static ../programs/lz4 -vqB4D | qemu-m68k-static ../programs/lz4 -qt
using blocks of size 64 KB

The futex facility returned an unexpected error code.
...
qemu: uncaught target signal 6 (Aborted) - core dumped
Aborted
lz4: lz4io.c:2192: LZ4IO_decompressLZ4F: Assertion `lastStatus == 0' failed.
qemu: uncaught target signal 6 (Aborted) - core dumped
Aborted
make[1]: *** [Makefile:310: test-platform] Error 134
```

</details>


But when I split piped commands into separated commands:

- Compression program (`lz4 -vqB4D`) aborts
  - But it generates proper `.lz4` file
- Decompression program (`lz4 -d`) decompress it properly
- Original file and decompressed file are same.

It may suggest that program/process itself is working correctly but it doesn't terminate something gracefully.

<details>

<summary>Same test, but splitted</summary>

```
cd ./tests/

qemu-m68k-static ./datagen -g256MB > data-256MB
# OK

qemu-m68k-static ../programs/lz4 -vqB4D data-256MB
using blocks of size 64 KB
Compressed filename will be : data-256MB.lz4
The futex facility returned an unexpected error code.
qemu: uncaught target signal 6 (Aborted) - core dumped
Aborted
# FAILED ?

qemu-m68k-static ../programs/lz4 -d data-256MB.lz4 data-256MB.lz4.dec
data-256MB.lz4                 : decoded 268435456 bytes
# OK

cmp data-256MB data-256MB.lz4.dec
# OK
```

</details>

--- a/programs/lz4io.c
+++ b/programs/lz4io.c
@@ -1129,24 +1129,17 @@ static size_t LZ4IO_compressFrameChunk(const void* params,
             END_PROCESS(51, "unable to create a LZ4F compression context");
     }
     /* init state, and writes frame header, will be overwritten at next stage. */
-    {   size_t const whr = LZ4F_compressBegin_usingCDict(cctx, dst, dstCapacity, cfcp->cdict, cfcp->prefs);
-        if (LZ4F_isError(whr))
-            END_PROCESS(52, "error initializing LZ4F compression context");
-    }
-    /* let's populate compression tables with Prefix */
     if (prefixSize) {
-        size_t pSize;
-        assert(cfcp->prefs->frameInfo.blockMode == LZ4F_blockLinked);
+        size_t const whr = LZ4F_compressBegin_usingDict(cctx, dst, dstCapacity, (const char*)src - prefixSize, prefixSize, cfcp->prefs);
+        if (LZ4F_isError(whr))
+            END_PROCESS(52, "error initializing LZ4F compression context with prefix");
         assert(prefixSize == 64 KB);
-        assert(srcSize >= prefixSize);
-        pSize = LZ4F_compressUpdate(cctx, dst, dstCapacity, (const char*)src - prefixSize, prefixSize, NULL);
-        if (LZ4F_isError(pSize))
-            END_PROCESS(53, "error compressing prefix");
-        pSize = LZ4F_flush(cctx, dst, dstCapacity, NULL);
-        if (LZ4F_isError(pSize))
-            END_PROCESS(54, "error flushing prefix");
+    } else {
+        size_t const whr = LZ4F_compressBegin_usingCDict(cctx, dst, dstCapacity, cfcp->cdict, cfcp->prefs);
+        if (LZ4F_isError(whr))
+            END_PROCESS(53, "error initializing LZ4F compression context");
     }
-    /* let's overwrite */
+    /* let's now compress, overwriting unused header */
     {   size_t const cSize = LZ4F_compressUpdate(cctx, dst, dstCapacity, src, srcSize, NULL);
         if (LZ4F_isError(cSize))
             END_PROCESS(55, "error compressing with LZ4F_compressUpdate");
@@ -1501,17 +1494,10 @@ LZ4IO_compressFilename_extRess(unsigned long long* inStreamSize,
                                int compressionLevel,
                                const LZ4IO_prefs_t* const io_prefs)
 {
-#if LZ4IO_MULTITHREAD
-    /* only employ multi-threading in the following scenarios: */
-    if ( (io_prefs->nbWorkers > 1)
-      && ( (io_prefs->blockIndependence == LZ4F_blockIndependent)  /* blocks must be independent */
-        || (!io_prefs->useDictionary) )
-      )
+    if (LZ4IO_MULTITHREAD)
         return LZ4IO_compressFilename_extRess_MT(inStreamSize, ress, srcFileName, dstFileName, compressionLevel, io_prefs);
-#endif
     /* Only single-thread available */
     return LZ4IO_compressFilename_extRess_ST(inStreamSize, ress, srcFileName, dstFileName, compressionLevel, io_prefs);
-
 }
 
 int LZ4IO_compressFilename(const char* srcFileName, const char* dstFileName, int compressionLevel, const LZ4IO_prefs_t* prefs)

--- a/programs/lz4io.c
+++ b/programs/lz4io.c
@@ -706,7 +706,8 @@ static void LZ4IO_readAndProcess(void* arg)
     if (!buffer)
         END_PROCESS(31, "Allocation error : can't allocate buffer to read new chunk");
     if (prefixSize) {
-        memcpy(buffer, rjd->prefix, 64 KB);
+        assert(prefixSize == 64 KB);
+        memcpy(buffer, rjd->prefix, prefixSize);
     }
     {   char* const in_buff = (char*)buffer + prefixSize;
         size_t const inSize = fread(in_buff, (size_t)1, chunkSize, rjd->fin);
@@ -744,7 +745,7 @@ static void LZ4IO_readAndProcess(void* arg)
             cjd->lastBlock = inSize < chunkSize;
             TPool_submitJob(rjd->tPool, LZ4IO_compressAndFreeChunk, cjd);
             if (inSize == chunkSize) {
-                /* probably more ? read another chunk */
+                /* likely more => read another chunk */
                 rjd->blockNb++;
                 TPool_submitJob(rjd->tPool, LZ4IO_readAndProcess, rjd);
     }   }   }
@@ -886,6 +887,7 @@ static int LZ4IO_compressLegacy_internal(unsigned long long* readSize,
                     (double)wr.totalCSize / (double)(rjd.totalReadSize + !rjd.totalReadSize) * 100.);
         *readSize = rjd.totalReadSize;
     }
+
     /* Close & Free */
 _cfl_clean:
     WR_destroy(&wr);
@@ -1126,18 +1128,30 @@ static size_t LZ4IO_compressFrameChunk(const void* params,
         if (cctx==NULL || LZ4F_isError(ccr))
             END_PROCESS(51, "unable to create a LZ4F compression context");
     }
-    /* init state, and writes frame header, will be overwritten at next stage.
-     * Also: no support for dictionary yet, meaning linked blocks are actually independent */
+    /* init state, and writes frame header, will be overwritten at next stage. */
     {   size_t const whr = LZ4F_compressBegin_usingCDict(cctx, dst, dstCapacity, cfcp->cdict, cfcp->prefs);
         if (LZ4F_isError(whr))
             END_PROCESS(52, "error initializing LZ4F compression context");
     }
+    /* let's populate compression tables with Prefix */
+    if (prefixSize) {
+        size_t pSize;
+        assert(cfcp->prefs->frameInfo.blockMode == LZ4F_blockLinked);
+        assert(prefixSize == 64 KB);
+        assert(srcSize >= prefixSize);
+        pSize = LZ4F_compressUpdate(cctx, dst, dstCapacity, (const char*)src - prefixSize, prefixSize, NULL);
+        if (LZ4F_isError(pSize))
+            END_PROCESS(53, "error compressing prefix");
+        pSize = LZ4F_flush(cctx, dst, dstCapacity, NULL);
+        if (LZ4F_isError(pSize))
+            END_PROCESS(54, "error flushing prefix");
+    }
+    /* let's overwrite */
     {   size_t const cSize = LZ4F_compressUpdate(cctx, dst, dstCapacity, src, srcSize, NULL);
         if (LZ4F_isError(cSize))
-            END_PROCESS(53, "error compressing with LZ4F_compressUpdate");
+            END_PROCESS(55, "error compressing with LZ4F_compressUpdate");
 
         LZ4F_freeCompressionContext(cctx);
-        (void)prefixSize;
         return (size_t) cSize;
     }
 }
@@ -1489,8 +1503,9 @@ LZ4IO_compressFilename_extRess(unsigned long long* inStreamSize,
 {
 #if LZ4IO_MULTITHREAD
     /* only employ multi-threading in the following scenarios: */
-    if ( (io_prefs->nbWorkers != 1)
-      && (io_prefs->blockIndependence == LZ4F_blockIndependent)  /* blocks must be independent */
+    if ( (io_prefs->nbWorkers > 1)
+      && ( (io_prefs->blockIndependence == LZ4F_blockIndependent)  /* blocks must be independent */
+        || (!io_prefs->useDictionary) )
       )
         return LZ4IO_compressFilename_extRess_MT(inStreamSize, ress, srcFileName, dstFileName, compressionLevel, io_prefs);
 #endif

--- a/programs/lz4io.c
+++ b/programs/lz4io.c
@@ -1163,7 +1163,7 @@ static size_t LZ4IO_compressFrameChunk(const void* params,
  */
 int
 LZ4IO_compressFilename_extRess_MT(unsigned long long* inStreamSize,
-                               cRess_t ress,
+                               cRess_t* ress,
                                const char* srcFileName, const char* dstFileName,
                                int compressionLevel,
                                const LZ4IO_prefs_t* const io_prefs)
@@ -1171,12 +1171,12 @@ LZ4IO_compressFilename_extRess_MT(unsigned long long* inStreamSize,
     unsigned long long filesize = 0;
     unsigned long long compressedfilesize = 0;
     FILE* dstFile;
-    void* const srcBuffer = ress.srcBuffer;
-    void* const dstBuffer = ress.dstBuffer;
-    const size_t dstBufferSize = ress.dstBufferSize;
+    void* const srcBuffer = ress->srcBuffer;
+    void* const dstBuffer = ress->dstBuffer;
+    const size_t dstBufferSize = ress->dstBufferSize;
     const size_t chunkSize = 4 MB;  /* each job should be "sufficiently large" */
     size_t readSize;
-    LZ4F_compressionContext_t ctx = ress.ctx;   /* just a pointer */
+    LZ4F_compressionContext_t ctx = ress->ctx;   /* just a pointer */
     LZ4F_preferences_t prefs;
 
     /* Init */
@@ -1186,7 +1186,7 @@ LZ4IO_compressFilename_extRess_MT(unsigned long long* inStreamSize,
     if (dstFile == NULL) { fclose(srcFile); return 1; }
 
     /* Adjust compression parameters */
-    prefs = ress.preparedPrefs;
+    prefs = ress->preparedPrefs;
     prefs.compressionLevel = compressionLevel;
     if (io_prefs->contentSizeFlag) {
       U64 const fileSize = UTIL_getOpenFileSize(srcFile);
@@ -1196,7 +1196,7 @@ LZ4IO_compressFilename_extRess_MT(unsigned long long* inStreamSize,
     }
 
     /* read first chunk */
-    assert(chunkSize <= ress.srcBufferSize);
+    assert(chunkSize <= ress->srcBufferSize);
     readSize  = fread(srcBuffer, (size_t)1, chunkSize, srcFile);
     if (ferror(srcFile))
         END_PROCESS(40, "Error reading first chunk (%u bytes) of '%s' ", (unsigned)chunkSize, srcFileName);
@@ -1205,7 +1205,7 @@ LZ4IO_compressFilename_extRess_MT(unsigned long long* inStreamSize,
     /* single-block file */
     if (readSize < chunkSize) {
         /* Compress in single pass */
-        size_t const cSize = LZ4F_compressFrame_usingCDict(ctx, dstBuffer, dstBufferSize, srcBuffer, readSize, ress.cdict, &prefs);
+        size_t const cSize = LZ4F_compressFrame_usingCDict(ctx, dstBuffer, dstBufferSize, srcBuffer, readSize, ress->cdict, &prefs);
         if (LZ4F_isError(cSize))
             END_PROCESS(41, "Compression failed : %s", LZ4F_getErrorName(cSize));
         compressedfilesize = cSize;
@@ -1229,17 +1229,17 @@ LZ4IO_compressFilename_extRess_MT(unsigned long long* inStreamSize,
         LZ4IO_CfcParameters cfcp;
         ReadTracker rjd;
 
-        if (ress.tPool == NULL) {
-            ress.tPool = TPool_create(io_prefs->nbWorkers, 4);
-            assert(ress.wPool == NULL);
-            ress.wPool = TPool_create(1, 4);
-            if (ress.tPool == NULL || ress.wPool == NULL)
+        if (ress->tPool == NULL) {
+            ress->tPool = TPool_create(io_prefs->nbWorkers, 4);
+            assert(ress->wPool == NULL);
+            ress->wPool = TPool_create(1, 4);
+            if (ress->tPool == NULL || ress->wPool == NULL)
                 END_PROCESS(43, "can't create threadpools");
         }
         cfcp.prefs = &prefs;
-        cfcp.cdict = ress.cdict;
-        rjd.tPool = ress.tPool;
-        rjd.wpool = ress.wPool;
+        cfcp.cdict = ress->cdict;
+        rjd.tPool = ress->tPool;
+        rjd.wpool = ress->wPool;
         rjd.fin = srcFile;
         rjd.chunkSize = chunkSize;
         rjd.totalReadSize = 0;
@@ -1285,7 +1285,7 @@ LZ4IO_compressFilename_extRess_MT(unsigned long long* inStreamSize,
 
         /* process first block */
         {   CompressJobDesc cjd;
-            cjd.wpool = ress.wPool;
+            cjd.wpool = ress->wPool;
             cjd.buffer = srcBuffer;
             cjd.prefixSize = 0;
             cjd.inSize = readSize;
@@ -1296,7 +1296,7 @@ LZ4IO_compressFilename_extRess_MT(unsigned long long* inStreamSize,
             cjd.wr = &wr;
             cjd.maxCBlockSize = rjd.maxCBlockSize;
             cjd.lastBlock = 0;
-            TPool_submitJob(ress.tPool, LZ4IO_compressChunk, &cjd);
+            TPool_submitJob(ress->tPool, LZ4IO_compressChunk, &cjd);
             rjd.totalReadSize = readSize;
             rjd.blockNb = 1;
             if (prefixBuffer) {
@@ -1305,11 +1305,11 @@ LZ4IO_compressFilename_extRess_MT(unsigned long long* inStreamSize,
             }
 
             /* Start the job chain */
-            TPool_submitJob(ress.tPool, LZ4IO_readAndProcess, &rjd);
+            TPool_submitJob(ress->tPool, LZ4IO_readAndProcess, &rjd);
 
             /* Wait for all completion */
-            TPool_jobsCompleted(ress.tPool);
-            TPool_jobsCompleted(ress.wPool);
+            TPool_jobsCompleted(ress->tPool);
+            TPool_jobsCompleted(ress->wPool);
             compressedfilesize += wr.totalCSize;
         }
 
@@ -1371,7 +1371,7 @@ LZ4IO_compressFilename_extRess_MT(unsigned long long* inStreamSize,
  */
 int
 LZ4IO_compressFilename_extRess_ST(unsigned long long* inStreamSize,
-                               cRess_t ress,
+                               const cRess_t* ress,
                                const char* srcFileName, const char* dstFileName,
                                int compressionLevel,
                                const LZ4IO_prefs_t* const io_prefs)
@@ -1379,12 +1379,12 @@ LZ4IO_compressFilename_extRess_ST(unsigned long long* inStreamSize,
     unsigned long long filesize = 0;
     unsigned long long compressedfilesize = 0;
     FILE* dstFile;
-    void* const srcBuffer = ress.srcBuffer;
-    void* const dstBuffer = ress.dstBuffer;
-    const size_t dstBufferSize = ress.dstBufferSize;
+    void* const srcBuffer = ress->srcBuffer;
+    void* const dstBuffer = ress->dstBuffer;
+    const size_t dstBufferSize = ress->dstBufferSize;
     const size_t blockSize = io_prefs->blockSize;
     size_t readSize;
-    LZ4F_compressionContext_t ctx = ress.ctx;   /* just a pointer */
+    LZ4F_compressionContext_t ctx = ress->ctx;   /* just a pointer */
     LZ4F_preferences_t prefs;
 
     /* Init */
@@ -1395,7 +1395,7 @@ LZ4IO_compressFilename_extRess_ST(unsigned long long* inStreamSize,
     memset(&prefs, 0, sizeof(prefs));
 
     /* Adjust compression parameters */
-    prefs = ress.preparedPrefs;
+    prefs = ress->preparedPrefs;
     prefs.compressionLevel = compressionLevel;
     if (io_prefs->contentSizeFlag) {
       U64 const fileSize = UTIL_getOpenFileSize(srcFile);
@@ -1412,7 +1412,7 @@ LZ4IO_compressFilename_extRess_ST(unsigned long long* inStreamSize,
     /* single-block file */
     if (readSize < blockSize) {
         /* Compress in single pass */
-        size_t const cSize = LZ4F_compressFrame_usingCDict(ctx, dstBuffer, dstBufferSize, srcBuffer, readSize, ress.cdict, &prefs);
+        size_t const cSize = LZ4F_compressFrame_usingCDict(ctx, dstBuffer, dstBufferSize, srcBuffer, readSize, ress->cdict, &prefs);
         if (LZ4F_isError(cSize))
             END_PROCESS(41, "Compression failed : %s", LZ4F_getErrorName(cSize));
         compressedfilesize = cSize;
@@ -1429,7 +1429,7 @@ LZ4IO_compressFilename_extRess_ST(unsigned long long* inStreamSize,
     /* multiple-blocks file */
     {
         /* Write Frame Header */
-        size_t const headerSize = LZ4F_compressBegin_usingCDict(ctx, dstBuffer, dstBufferSize, ress.cdict, &prefs);
+        size_t const headerSize = LZ4F_compressBegin_usingCDict(ctx, dstBuffer, dstBufferSize, ress->cdict, &prefs);
         if (LZ4F_isError(headerSize))
             END_PROCESS(43, "File header generation failed : %s", LZ4F_getErrorName(headerSize));
         if (fwrite(dstBuffer, 1, headerSize, dstFile) != headerSize)
@@ -1496,7 +1496,7 @@ LZ4IO_compressFilename_extRess_ST(unsigned long long* inStreamSize,
 
 static int
 LZ4IO_compressFilename_extRess(unsigned long long* inStreamSize,
-                               cRess_t ress,
+                               cRess_t* ress,
                                const char* srcFileName, const char* dstFileName,
                                int compressionLevel,
                                const LZ4IO_prefs_t* const io_prefs)
@@ -1518,10 +1518,10 @@ int LZ4IO_compressFilename(const char* srcFileName, const char* dstFileName, int
 {
     TIME_t const timeStart = TIME_getTime();
     clock_t const cpuStart = clock();
-    cRess_t const ress = LZ4IO_createCResources(prefs);
+    cRess_t ress = LZ4IO_createCResources(prefs);
     unsigned long long processed;
 
-    int const result = LZ4IO_compressFilename_extRess(&processed, ress, srcFileName, dstFileName, compressionLevel, prefs);
+    int const result = LZ4IO_compressFilename_extRess(&processed, &ress, srcFileName, dstFileName, compressionLevel, prefs);
 
     /* Free resources */
     LZ4IO_freeCResources(ress);
@@ -1556,7 +1556,7 @@ int LZ4IO_compressMultipleFilenames(
         unsigned long long processed;
         size_t const ifnSize = strlen(inFileNamesTable[i]);
         if (LZ4IO_isStdout(suffix)) {
-            missed_files += LZ4IO_compressFilename_extRess(&processed, ress,
+            missed_files += LZ4IO_compressFilename_extRess(&processed, &ress,
                                     inFileNamesTable[i], stdoutmark,
                                     compressionLevel, prefs);
             totalProcessed += processed;
@@ -1574,7 +1574,7 @@ int LZ4IO_compressMultipleFilenames(
         strcpy(dstFileName, inFileNamesTable[i]);
         strcat(dstFileName, suffix);
 
-        missed_files += LZ4IO_compressFilename_extRess(&processed, ress,
+        missed_files += LZ4IO_compressFilename_extRess(&processed, &ress,
                                 inFileNamesTable[i], dstFileName,
                                 compressionLevel, prefs);
         totalProcessed += processed;

